docs/architecture-decisions/2024-12-10_arrays.md

```diff
@@ -0,0 +1,237 @@
+-   **Date created**: 2024-12-10
+-   **Date decided**:
+-   **Date updated**:
+
+## Context
+
+Arrays are a fundamental primitive in a high level languages and as such broad support is desirable, even in a constrained environment like the AVM. Currently Algorand Python only supports ARC-4 encoded arrays that have been designed to be efficient for reading, letting clients interacting with smart contracts manage the additional work required to ARC-4 encode the values. Given they are not efficient for creating or manipulating values, we have room to improve and introduce a more general purpose array.
+
+In addition to how an array is encoded, the other facet to consider is where the array is stored in memory. The AVM supports two different types of memory which can be leveraged when writing a program: the stack and scratch space. These concepts are akin to a stack and heap found in more traditional virtual machine environments like the JVM.
+
+The AVM memory types have some constraints that need to be considered. The stack can hold up to 1000 values, but only 256 are addressable at any point. Scratch slots support up to 256 addressable values. Both support values types of either a single uint64 or a byte slice of up to 4096 bytes. The stack is addressed using relative static offsets whereas scratch slots are addressed using dynamic absolute indexes. The stack is a good fit for value based semantics as each value is a copy, whereas scratch slots are a good fit for reference based semantics as each value can be uniquely addressed.
+
+In our target high level languages (Python and JavaScript), the builtin arrays (or lists) types have reference based semantics and mutable API's. Meaning that when an array is modified, all references to that array see the modification.
+
+The developer experience with Puya Arrays will depend on the API design (mutable vs immutable) and where those arrays reside in memory (stack vs scratch) this ADR will outline the pros and cons of some different approaches with these facets in mind.
+
+## Requirements
+
+-   Provide a more efficient array implementation for general purpose use
+-   Provide an API that communicates clearly the effect of operations and is semantically correct
+-   Support a broad range of types (including non ARC-4 types)
+
+## Principles
+
+-   [AlgoKit Guiding Principles](https://github.com/algorandfoundation/algokit-cli/blob/main/docs/algokit.md#guiding-principles) - specifically Seamless onramp, Meet developers where they are, Leverage existing ecosystem
+-   [Algorand Python Principles](https://algorandfoundation.github.io/puya/principles.html#principles) - specifically Least surprise and Leverage existing ecosystem by maintaining semantic compatibility.
+-   [Algorand TypeScript Guiding Principles](https://github.com/algorandfoundation/puya-ts/blob/main/docs/README.md#guiding-principals)
```

Comment:
Thanks for putting this together @neilcampbell. This looks like it captured our discussions, options, and decision well. This is more of a comment re: the Algorand Typescript Guiding Principles. Similar to the Python guiding principles, there should probably a principle about maintaining semantic compatibility with Typescript.

Reply:
I'm happy to make this clearer. I guess the intent of the "Algorand Python" section within the guiding principles is to inherit the Algorand Python guiding principles and reduce duplication.